Handbook: New best practices for shipping and announcing features (#7485)
PostHog/posthog.com: contents/handbook/engineering/development-process.md

---

> title: Shipping & Releasing

Reviewer: lowercase "releasing"! :-P

Reviewer: Also, should the title more closely match the slug?

Author: ```suggestion
title: Shipping & releasing
```

sidebar: Handbook
showTitle: true
---

Any process is a balance between speed and control. If we have a long process that requires extensive QA and 10 approvals, we will never make mistakes because we will never release anything.

However, if we have no checks in place, we will release quickly but everything will be broken. For this reason we have some best practices for releasing things, and guidelines on how to ship.

## How to decide what to build

### Set milestones

To start, Product and Engineering should align on major milestones (e.g. Collaboration) which we have strong conviction will drive our success metrics for a feature.

There are two types of goals.

> * **Moonshots:** These are big, scary goals where we expect to fail 50% of the time. If we fail we expect to learn something equally as valuable as if we succeed. Just scraping the goal counts as a success.

Reviewer 2: I know this wasn't your writing so feel free to ignore for this PR, but do we really go for moonshots, given the individual products we're building already have product-market fit? Even data warehouse, which is a very ambitious product, I'd consider a roofshot given everyone knows what data warehouses are, they already exist etc.

Author: I defer to the VP of Engineering, aka @benjackwhite

Reviewer: Narrator: But PostHog doesn't actually do moonshots.

Reviewer 2: Let's merge this in now with Joe's edits and then spin up a separate PR on the whole moonshots thing, but yeah I agree it doesn't jive with our actual product strategy.

### Break up goals

The owner turns the ambiguous milestone into a roadmap of ambitious, meaningful, sprint-sized goals, thinking 2 - 3 sprints ahead to give other functions time. [Goal principles](/handbook/engineering/development-process#set-some-milestones) still apply.

### Execute on goals

Our sprint planning session is where the high-level goals for the next sprint get decided. Each team takes these, breaks them down, and adds their own smaller items, like bugfixes.

Sprint planning works like this:
* During the week of the sprint planning, teams (with the help of the product managers) define one or two big ambitious goals for this sprint that align with major milestones.
* On Wednesday we get together. We use breakout rooms to split into the small teams, and each team finalizes the ambitious goals that they want to tackle.
* We come back in the same room, where
* We do a quick retrospective on the goals for the last sprint (did we get everything done? Should we have prioritized other things?)

> * Each team lead presents the main goals for their team. We have an open discussion about whether the goals are the right goals for this sprint and whether they are ambitious enough.

Reviewer 3: We haven't done that in a long time, teams just do their planning as part of their own meetings (or within the Wednesday standup), then the engineering-wide meeting is just a way of collecting the goals of all teams.
So of these four points, only the "Each team lead presents" one is up to date.

Author: I'm not deeply familiar with Engineering team processes, so I'm going to defer to Engineering team to keep this aspect up to date tbh.

### Evaluate success 

Review impact of each major milestone and feedback into the planning process.

When we review the status of goals we classify them as follows:

* **Nailed it:** We hit the goal spectacularly. High fives all round.
* **Scraped it:** We _almost_ hit the goal, but we'll need to do a little bit more next sprint to tidy up. We should adjust our workload to have fewer resources on big goals during the next sprint to comfortable get this finished.
* **Failed it:** We were nowhere near hitting the goal, but we learned some valuable lessons. We're going to go back to the drawing board. Maybe the goal wasn't right or maybe there's a different way to approach it?

## What about the small stuff?

Not everything directly contributes to a company level goal. It’s important that the small stuff also gets done for us to succeed. Use the following principles:

* **Yes, and**: Be encouraging and helpful with others who are innovating. All of our biggest wins have looked like bad ideas early on.
* **Dogfooding**: Use the product yourself. When you see something that annoys you, fix it.

> * **Sidequests:** Smaller projects you are passionate about but may not shoot up our metrics (e.g. turbo mode).

Reviewer: Sidequests → Side quests

Author: ```suggestion
* **Side quests:** Smaller projects you are passionate about but may not shoot up our metrics (e.g. turbo mode).
```

* **Support hero:** Support hero dedicates all of their time to customers, solving the wild and wonderful issues our customers find each week.

## Sizing tasks and reducing WIP

Efficient engineering organizations actively [reduce Work In Progress](https://loom.com/share/5efceb288b634a449041918bdba08202) (WIP) to avoid negative feedback loops that drive down productivity. 

> Hence, a PR should be optimised for two things: 

Reviewer: US not EU!! ;) 

optimised → optimized

Author: ```suggestion
Hence, a PR should be optimized for two things: 
```

> 1. Quality of implementation.

Reviewer: No periods after fragments ;)

> 1. Quality of implementation.
> 2. The speed with which we can merge it in.

Author: ```suggestion
1. Quality of implementation
2. The speed with which we can merge it in
```
